image_segmentation/pytorch/runtime/distributed_utils.py

```diff
@@ -0,0 +1,145 @@
+    random.seed(seed)
+
+
+def generate_seeds(rng, size):
+    """
+    Generate list of random seeds
+
+    :param rng: random number generator
+    :param size: length of the returned list
+    """
+    seeds = [rng.randint(0, 2**32 - 1) for _ in range(size)]
+    return seeds
+
+
+def broadcast_seeds(seeds, device):
+    """
+    Broadcasts random seeds to all distributed workers.
+    Returns list of random seeds (broadcasted from workers with rank 0).
+
+    :param seeds: list of seeds (integers)
+    :param device: torch.device
+    """
+    if torch.distributed.is_available() and torch.distributed.is_initialized():
+        seeds_tensor = torch.LongTensor(seeds).to(device)
+        torch.distributed.broadcast(seeds_tensor, 0)
+        seeds = seeds_tensor.tolist()
+    return seeds
+
+
+def setup_seeds(master_seed, epochs, device):
+    """
+    Generates seeds from one master_seed.
+    Function returns (worker_seeds, shuffling_seeds), worker_seeds are later
+    used to initialize per-worker random number generators (mostly for
+    dropouts), shuffling_seeds are for RNGs resposible for reshuffling the
+    dataset before each epoch.
+    Seeds are generated on worker with rank 0 and broadcasted to all other
+    workers.
+
+    :param master_seed: master RNG seed used to initialize other generators
+    :param epochs: number of epochs
+    :param device: torch.device (used for distributed.broadcast)
+    """
+    if master_seed == -1:
+        # random master seed, random.SystemRandom() uses /dev/urandom on Unix
+        master_seed = random.SystemRandom().randint(0, 2**32 - 1)
+        if get_rank() == 0:
+            # master seed is reported only from rank=0 worker, it's to avoid
+            # confusion, seeds from rank=0 are later broadcasted to other
+            # workers
+            print(f'Using random master seed: {master_seed}')
+    else:
+        # master seed was specified from command line
+        print(f'Using master seed from command line: {master_seed}')
+
+    # initialize seeding RNG
+    seeding_rng = random.Random(master_seed)
+
+    # generate worker seeds, one seed for every distributed worker
+    worker_seeds = generate_seeds(seeding_rng, get_world_size())
+
+    # generate seeds for data shuffling, one seed for every epoch
+    shuffling_seeds = generate_seeds(seeding_rng, epochs)
+
+    # broadcast seeds from rank=0 to other workers
+    worker_seeds = broadcast_seeds(worker_seeds, device)
+    shuffling_seeds = broadcast_seeds(shuffling_seeds, device)
+    return worker_seeds, shuffling_seeds
+
+
+def get_world_size():
+    return int(os.environ.get('WORLD_SIZE', 1))
+
+
+def reduce_tensor(tensor, num_gpus):
+    if num_gpus > 1:
+        rt = tensor.clone()
+        dist.all_reduce(rt, op=dist.reduce_op.SUM)
+        if rt.is_floating_point():
+            rt = rt / num_gpus
+        else:
+            rt = rt // num_gpus
+        return rt
+    return tensor
+
+
+def init_distributed():
```

Comment:
Are we allowed the do data parallel?

Reply:
Well, technically yes, but that would require rewriting the training function completely. Will you train with DP or that is a hypothetical question? If you have a concrete usecase we can surely talk about it.

Technically, multi-worker training functionality is not a must-have in a reference implementation. I did it with DDP to accelerate the experiments, but the final results are always done using a single accelerator.

Reply:
It's a hypothetical question. I was thinking that the benchmark rules specified single node training so I wasn't sure if there was an actual DP case or if this was just future looking.

Reply:
This is an additional (optional) feature that I used to get the preliminary results faster. I decided to leave it in the code since you might find it handy.